File: android/wearscope/src/main/kotlin/io/wearscope/WearScope.kt
/*
 * WearScope — observability SDK for smart-glasses (DAT) apps (Kotlin).
 *
 *   WearScope.start(context, apiKey = "ws_dev")                    // local mode (file logging only)
 *   WearScope.start(context, apiKey = "gs_...", endpoint = "https://...")
 *   WearScope.track(WSEventType.PHOTO, "capture", mapOf("ms" to "401"))
 *   WearScope.trackError(e, context = "camera.session")            // known DAT errors get an explanation attached
 *   WearScope.observeAudioRoutes(context)                          // glasses↔phone route changes
 */
package io.wearscope

import android.content.Context
import android.util.Log
import android.media.AudioDeviceCallback
import android.media.AudioDeviceInfo
import android.media.AudioManager
import java.io.File
import java.util.Locale
import java.util.TimeZone

object WearScope {
  const val SDK_VERSION = "0.4.0"

  /** Start the SDK. A null endpoint means local mode (events go to file only). */
  @JvmStatic
  fun start(context: Context, apiKey: String, endpoint: String? = null) {
    WSCore.start(context.applicationContext, apiKey, endpoint)
    track(WSEventType.CUSTOM, "sdk_start", mapOf(
        "sdk" to SDK_VERSION,
        "mode" to if (endpoint == null) "local" else "cloud",
        "locale" to Locale.getDefault().toString(),
        "timezone" to TimeZone.getDefault().id,
    ))
  }

  /**
   * Zero-config start — no API key, no signup.
   *
   * First launch asks the cloud for an anonymous project, caches the credentials,
   * and logs the dashboard URL once. Later launches reuse the cached key. If the
   * server is unreachable the SDK runs in local mode and retries next launch, so
   * instrumentation never blocks or crashes an app.
   *
   * Pass [endpoint] to self-host. Data is claimable into an account later.
   */
  @JvmStatic
  @JvmOverloads
  fun start(context: Context, endpoint: String? = null) {
    val app = context.applicationContext
    val target = endpoint ?: WSProvisioning.DEFAULT_ENDPOINT
    WSProvisioning.cached(app)?.let { cached ->
      WSProvisioning.announce(cached, fresh = false)
      start(app, cached.apiKey, target)
      return
    }
    // Buffer locally from the very first event; upload begins once a key exists.
    start(app, "", null)
    Thread {
      val name = runCatching {
        app.packageManager.getApplicationLabel(app.applicationInfo).toString()
      }.getOrDefault(app.packageName)
      val creds = WSProvisioning.provision(app, target, name)
      if (creds == null) {
        Log.i("WearScope", "Cloud unavailable — local mode (events kept on device)")
      } else {
        WSProvisioning.announce(creds, fresh = true)
        WSCore.adopt(creds.apiKey, target)
      }
    }.apply { isDaemon = true }.start()
  }

  /** Record an event. attrs is metadata only (no payloads — privacy principle). */
  @JvmStatic
  fun track(type: WSEventType, name: String, attrs: Map<String, String> = emptyMap()) {
    WSCore.add(WSEvent(type, name, attrs))
  }

  /** Record an error — known DAT failure modes get an explanation attached automatically. */
  @JvmStatic
  fun trackError(description: String, context: String) {
    val attrs = buildMap {
      put("context", context)
      put("description", description.take(500))
      WSErrorDecoder.explain(description)?.let { put("explain", it) }
    }
    WSCore.add(WSEvent(WSEventType.ERROR, context, attrs))
  }

  @JvmStatic
  fun trackError(throwable: Throwable, context: String) =
      trackError(throwable.toString(), context)

  /** Duration helper: call the returned lambda on completion to record elapsed ms. */
  @JvmStatic
  fun measure(
      type: WSEventType,
      name: String,
      attrs: Map<String, String> = emptyMap(),
  ): (Map<String, String>) -> Unit {
    val t0 = System.currentTimeMillis()
    return { extra ->
      track(type, name, attrs + extra + ("ms" to (System.currentTimeMillis() - t0).toString()))
    }
  }

  /** Attempt an immediate upload (cloud mode). */
  @JvmStatic fun flush() = WSCore.flush()

  /** Local event file (jsonl) — for debugging and sharing. */
  @JvmStatic fun exportFile(): File? = WSCore.exportFile()

  // MARK: - Audio route observation (detects the silent glasses↔phone mic fallback)

  /** Records one snapshot of the current route at start, then every device add/remove after. */
  @JvmStatic
  fun observeAudioRoutes(context: Context) {
    val am = context.getSystemService(Context.AUDIO_SERVICE) as AudioManager
    trackRoutes(am, reason = "initial")
    am.registerAudioDeviceCallback(object : AudioDeviceCallback() {
      override fun onAudioDevicesAdded(added: Array<out AudioDeviceInfo>) =
          trackRoutes(am, reason = "added:" + added.joinToString(",") { describe(it) })

      override fun onAudioDevicesRemoved(removed: Array<out AudioDeviceInfo>) =
          trackRoutes(am, reason = "removed:" + removed.joinToString(",") { describe(it) })
    }, null)
  }

  private fun trackRoutes(am: AudioManager, reason: String) {
    val inputs = am.getDevices(AudioManager.GET_DEVICES_INPUTS).joinToString(",") { describe(it) }
    val outputs = am.getDevices(AudioManager.GET_DEVICES_OUTPUTS)
        .filter { it.type != AudioDeviceInfo.TYPE_TELEPHONY }
        .joinToString(",") { describe(it) }
    track(WSEventType.AUDIO_ROUTE, "route", mapOf(
        "reason" to reason.take(200),
        "inputs" to inputs.take(300),   // glasses show as BT_SCO(device name) — the fallback-detection key
        "outputs" to outputs.take(300),
    ))
  }

  private fun describe(device: AudioDeviceInfo): String {
    val type = when (device.type) {
      AudioDeviceInfo.TYPE_BLUETOOTH_SCO -> "BT_SCO"
      AudioDeviceInfo.TYPE_BLUETOOTH_A2DP -> "BT_A2DP"
      AudioDeviceInfo.TYPE_BLE_HEADSET -> "BLE_HEADSET"
      AudioDeviceInfo.TYPE_BUILTIN_MIC -> "BUILTIN_MIC"
      AudioDeviceInfo.TYPE_BUILTIN_SPEAKER -> "SPEAKER"
      AudioDeviceInfo.TYPE_BUILTIN_EARPIECE -> "EARPIECE"
      else -> "T${device.type}"
    }
    return "$type(${device.productName})"
  }
}

File: Sources/WearScope/WearScope.swift
/*
 * WearScope — observability SDK for smart-glasses (DAT) apps. docs/OBSERVABILITY.md
 *
 * Usage:
 *   WearScope.start(apiKey: "ws_dev")                       // local mode (file logging only)
 *   WearScope.start(apiKey: "gs_...", endpoint: serverURL)  // cloud upload
 *   WearScope.track(.photo, "capture", ["ms": "401", "bytes": "48000"])
 *   WearScope.trackError(error, context: "hud.send")        // known DAT errors get an explanation attached
 */

import Foundation
#if canImport(UIKit)
import UIKit
#endif
#if canImport(AVFAudio)
import AVFAudio
#endif

public struct WSConfig: Sendable {
  public static let sdkVersion = "0.4.0"

  let apiKey: String
  let endpoint: URL?
  let appInfo: WSEnvelope.App
  let deviceInfo: WSEnvelope.Device

  init(apiKey: String, endpoint: URL?) {
    self.apiKey = apiKey
    self.endpoint = endpoint
    let bundle = Bundle.main
    appInfo = .init(
      bundleId: bundle.bundleIdentifier ?? "unknown",
      version: bundle.object(forInfoDictionaryKey: "CFBundleShortVersionString") as? String ?? "0",
      build: bundle.object(forInfoDictionaryKey: "CFBundleVersion") as? String ?? "0")
    #if canImport(UIKit)
    var model = ""
    var systemInfo = utsname()
    uname(&systemInfo)
    withUnsafeBytes(of: &systemInfo.machine) { bytes in
      model = String(decoding: bytes.prefix(while: { $0 != 0 }), as: UTF8.self)
    }
    deviceInfo = .init(model: model, os: "iOS \(UIDevice.current.systemVersion)")
    #else
    deviceInfo = .init(model: "unknown", os: "unknown")
    #endif
  }
}

public enum WearScope {

  /// Start the SDK. A nil endpoint means local mode (events go to file only).
  public static func start(apiKey: String, endpoint: URL? = nil) {
    let config = WSConfig(apiKey: apiKey, endpoint: endpoint)
    Task { await WSCore.shared.start(config: config) }
  }

  /// Record an event. attrs is metadata only (no payloads — privacy principle).
  public static func track(_ type: WSEventType, _ name: String, _ attrs: [String: String] = [:]) {
    Task { await WSCore.shared.add(WSEvent(type: type, name: name, attrs: attrs)) }
  }

  /// Record an error — known DAT failure modes get an explanation attached automatically.
  public static func trackError(_ error: Error, context: String) {
    trackError("\(error)", context: context)
  }

  public static func trackError(_ description: String, context: String) {
    var attrs = ["context": context, "description": String(description.prefix(500))]
    if let explanation = WSErrorDecoder.explain(description) {
      attrs["explain"] = explanation
    }
    Task { await WSCore.shared.add(WSEvent(type: .error, name: context, attrs: attrs)) }
  }

  /// Duration helper: call the returned closure on completion to record elapsed ms.
  public static func measure(_ type: WSEventType, _ name: String,
                             _ attrs: [String: String] = [:]) -> (@Sendable ([String: String]) -> Void) {
    let t0 = Date()
    return { extra in
      var merged = attrs.merging(extra) { _, new in new }
      merged["ms"] = String(Int(Date().timeIntervalSince(t0) * 1000))
      track(type, name, merged)
    }
  }

  #if canImport(AVFAudio) && os(iOS)
  /// Auto-observe audio routes — built to catch the silent glasses(HFP)↔phone mic fallback.
  /// Records one snapshot of the current route at start, then every change after.
  public static func observeAudioRoutes() {
    trackCurrentRoute(reason: "initial")
    NotificationCenter.default.addObserver(
      forName: AVAudioSession.routeChangeNotification, object: nil, queue: .main
    ) { note in
      let raw = (note.userInfo?[AVAudioSessionRouteChangeReasonKey] as? UInt) ?? 0
      let reason = AVAudioSession.RouteChangeReason(rawValue: raw)
      trackCurrentRoute(reason: routeReasonName(reason))
    }
  }

  private static func trackCurrentRoute(reason: String) {
    let route = AVAudioSession.sharedInstance().currentRoute
    let fmt: ([AVAudioSessionPortDescription]) -> String = { ports in
      ports.map { "\($0.portType.rawValue)(\($0.portName))" }.joined(separator: ",")
    }
    track(.audioRoute, "route", [
      "reason": reason,
      "inputs": fmt(route.inputs),    // glasses show as BluetoothHFP(device name) — the fallback-detection key
      "outputs": fmt(route.outputs),
    ])
  }

  private static func routeReasonName(_ r: AVAudioSession.RouteChangeReason?) -> String {
    switch r {
    case .newDeviceAvailable: return "newDeviceAvailable"
    case .oldDeviceUnavailable: return "oldDeviceUnavailable"
    case .categoryChange: return "categoryChange"
    case .override: return "override"
    case .wakeFromSleep: return "wakeFromSleep"
    case .noSuitableRouteForCategory: return "noSuitableRouteForCategory"
    case .routeConfigurationChange: return "routeConfigurationChange"
    default: return "unknown"
    }
  }
  #endif

  /// Attempt an immediate upload (cloud mode).
  public static func flush() {
    Task { await WSCore.shared.flush() }
  }

  /// Local event file (jsonl) — for debugging and sharing.
  public static func exportURL() async -> URL? {
    await WSCore.shared.localFileURL
  }
}
